File: web/src/api/bmsync.ts
import { api } from './index'

export interface SyncNode {
  syncId: string
  type: 'folder' | 'bookmark'
  title: string
  url?: string
  parentSyncId?: string
  index: number
  createdAt: string
  updatedAt: string
  deletedAt?: string
  rev?: number
}

export interface BmsyncStatus {
  configured: boolean
  serverUrl: string
  hasSynced: boolean
  rev: number
}

export interface ChangeIntent {
  op: 'create' | 'update' | 'delete'
  syncId?: string
  type?: 'folder' | 'bookmark'
  title?: string
  url?: string
  parentSyncId?: string
  index?: number
}

export async function getStatus(): Promise<BmsyncStatus> {
  const { data } = await api.get('/bmsync/status')
  return data
}

export async function getTree(): Promise<{ rev: number; nodes: SyncNode[] }> {
  const { data } = await api.get('/bmsync/tree')
  return data
}

export async function pullTree(): Promise<{ rev: number; nodes: SyncNode[] }> {
  const { data } = await api.post('/bmsync/pull')
  return data
}

export async function pushChanges(changes: ChangeIntent[]): Promise<{ rev: number; nodes: SyncNode[] }> {
  const { data } = await api.post('/bmsync/push', { changes })
  return data
}

File: server/models/bmsync.go
package models

// SyncNode mirrors the canonical bookmark node from the bookmark-sync server.
// Field names match the wire protocol exactly (camelCase JSON).
type SyncNode struct {
	SyncID       string `json:"syncId"`
	Type         string `json:"type"` // "bookmark" | "folder"
	Title        string `json:"title"`
	URL          string `json:"url,omitempty"`
	ParentSyncID string `json:"parentSyncId,omitempty"`
	Index        int    `json:"index"`
	CreatedAt    string `json:"createdAt"`
	UpdatedAt    string `json:"updatedAt"`
	DeletedAt    string `json:"deletedAt,omitempty"` // tombstone
	Rev          int    `json:"rev,omitempty"`
}

// SyncNodeResponse is the wire shape returned by the bookmark-sync server
// (POST /api/sync and GET /api/state): full canonical state + monotonic rev.
type SyncNodeResponse struct {
	Rev   int         `json:"rev"`
	Nodes []*SyncNode `json:"nodes"`
}

// ChangeIntent is a user mutation coming from the sundash UI.
// Op is one of "create" | "update" | "delete".
type ChangeIntent struct {
	Op           string `json:"op" binding:"required"`
	SyncID       string `json:"syncId"`
	Type         string `json:"type"`
	Title        string `json:"title"`
	URL          string `json:"url,omitempty"`
	ParentSyncID string `json:"parentSyncId,omitempty"`
	Index        int    `json:"index"`
}
